### PR TITLE
[builder] only remove subdirectories from META-INF/maven dir during `-pom`

### PR DESCRIPTION
**Pull request by rotty3000**

### CODE REVIEW

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java [hunk 1]
```diff
@@ -1210,4 +1210,14 @@ public void removePrefix(String prefixLow) {
...
 			.clear();
 	}
+
+	public void removeSubDirs(String prefixLow) {
+		String prefixHigh = prefixLow + "\uFFFF";
+		if (prefixLow.endsWith("/")) {
+			prefixLow = prefixLow.substring(0, prefixLow.length() - 1);
+			prefixHigh = prefixLow + "\uFFFF";
+		}
+		directories.subMap(prefixLow, prefixHigh)
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java]: This seems wrong. You always must modify both resources and directories. It is wrong to only clear things out of directories and not resources. You also need to clear the resources in those directories from resources.

Comment (rotty3000) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java]: yeah, you're right. My previous logic was better and did `removePrefx()` on each subdir dir. I'll re-revert to that. thx

Comment (rotty3000) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java]: Updated

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java [hunk 2]
```diff
@@ -1210,4 +1210,24 @@ public void removePrefix(String prefixLow) {
...
+		}
+
+		Map<String, Resource> dir = directories.get(path);
+
+		if ((dir == null) || dir.isEmpty()) {
+			return;
+		}
+
+		dir.keySet()
+			.forEach(s -> {
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java]: Can't this be simplified to 

```java
dir.keySet().forEach(this::removePrefix);
```

I cannot imagine the null check is needed and emptiness would be handled anyway.

Comment (rotty3000) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java]: the `isEmpty` is not needed, but the null check is.

Comment (bjhargrave) [reply biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java]: Why would the keySet have null values in it?

**Comment by rotty3000** [comment]
After much turbulence I've got a fix and some tests.

**Comment by rotty3000** [comment]
ping @bjhargrave

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java [hunk 3]
```diff
@@ -1210,4 +1210,28 @@ public void removePrefix(String prefixLow) {
 		directories.subMap(prefixLow, prefixHigh)
 			.clear();
 	}
+
+	public void removeSubDirs(String prefixLow) {
```
Comment (bjhargrave) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java]: Here is a much simpler implementation which I tested using the test cases here.

```java
	public void removeSubDirs(String prefixLow) {
		if (prefixLow.endsWith("/")) {
			prefixLow = prefixLow.substring(0, prefixLow.length() - 1);
		}
		Set<String> subDirs = new HashSet<>(directories.subMap(prefixLow, prefixLow + "\uFFFF")
			.keySet());
		subDirs.remove(prefixLow);
		subDirs.forEach(subDir -> removePrefix(subDir + "/"));
	}
```

**Comment by bjhargrave** [comment]
There is a better fix in my edited comment:

```java
	public void removeSubDirs(String dir) {
		if (!dir.endsWith("/")) {
			dir = dir + "/";
		}
		List<String> subDirs = new ArrayList<>(directories.subMap(dir, dir + "\uFFFF")
			.keySet());
		subDirs.forEach(subDir -> removePrefix(subDir + "/"));
	}
```

Please use it.

**Comment by rotty3000** [comment]
re-updated...

**Pull request merged**